lib/tests/streamlit/beta_util_test.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2018-2021 Streamlit Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+from unittest.mock import patch
+
+from streamlit.beta_util import function_beta_warning, object_beta_warning
+
+
+class BetaUtilTest(unittest.TestCase):
+    @patch("streamlit.warning")
+    def test_function_beta_warning(self, mock_warning):
+        def multiply(a, b):
+            return a * b
+
+        beta_multiply = function_beta_warning(multiply, "1980-01-01")
+
+        self.assertEqual(beta_multiply(3, 2), 6)
+        mock_warning.assert_called_once_with(
+            "`st.multiply` has graduated out of beta. On 1980-01-01, the beta_ version will be removed."
+            "\n\nBefore then, update your code from `st.beta_multiply` to `st.multiply`."
+        )
+
+    @patch("streamlit.warning")
+    def test_object_beta_warning(self, mock_warning):
+        class Multiplier(dict):
+            def multiply(self, a, b):
+                return a * b
+
+        beta_multiplier = object_beta_warning(Multiplier(), "multiplier", "1980-01-01")
+
+        expected_warning = (
+            "`st.multiplier` has graduated out of beta. On 1980-01-01, the beta_ version will be removed."
+            "\n\nBefore then, update your code from `st.beta_multiplier` to `st.multiplier`."
+        )
+
+        mock_warning.reset_mock()
+        self.assertEqual(beta_multiplier.multiply(3, 2), 6)
+        mock_warning.assert_called_once_with(expected_warning)
+
+        # Test that we also override various dunder methods.
+        # Some of these end up emitting the warning multiple times.
+        mock_warning.reset_mock()
+        beta_multiplier["foo"] = "bar"
+        mock_warning.assert_called_with(expected_warning)
```

Comment:
Should be `assert_called_once_with` right if we reset?

I ask cause I see we override `__getattr__`

Reply:
For what it's worth, I updated it with the `once` logical, and got an error.

```
AssertionError: Expected 'warning' to be called once. Called 3 times.
E           Calls: [call('`st.multiplier` has graduated out of beta. On 1980-01-01, the beta_ version will be removed.\n\nBefore then, update your code from `st.beta_multiplier` to `st.multiplier`.'),
E            call('`st.multiplier` has graduated out of beta. On 1980-01-01, the beta_ version will be removed.\n\nBefore then, update your code from `st.beta_multiplier` to `st.multiplier`.'),
E            call('`st.multiplier` has graduated out of beta. On 1980-01-01, the beta_ version will be removed.\n\nBefore then, update your code from `st.beta_multiplier` to `st.multiplier`.')].
```

We might be doing too much.

Reply:
There's a comment above ("some of these end up emitting the warning multiple times"). This happens because some of those functions are called multiple times during, e.g, iteration of the dictionary.

I don't think this is an issue - we're talking about a very very very niche case.

It may be that what we want is to only ever emit a single beta warning per script run. (This isn't what the previous code did, but I can see an argument for it with more complex cases!) But at the same time, this "advanced beta_warning logic" has basically become its own little feature...

Reply:
As of the latest commit, the warning should only ever be shown once for a single object, rather than every time one of that object's methods is called.